zrml/swaps/src/arbitrage.rs

```diff
@@ -0,0 +1,423 @@
+
+use crate::{math::calc_spot_price, root::calc_preimage};
+use alloc::collections::btree_map::BTreeMap;
+use parity_scale_codec::MaxEncodedLen;
+use sp_runtime::{
+    traits::{AtLeast32Bit, AtLeast32BitUnsigned},
+    SaturatedConversion,
+};
+use zeitgeist_primitives::{
+    constants::BASE,
+    types::{Asset, Pool},
+};
+
+// TODO  Why do we need the `Fixed`/`u128` type to begin with? Can't we just use a generic
+// `Balance` for all Balancer math functions?
+type Fixed = u128;
+
+const TOLERANCE: Fixed = BASE / 1_000; // 0.001
+
+/// This trait implements approximations for on-chain arbitrage of CPMM pools using bisection.
+///
+/// All calculations depend on on-chain, which are passed using the `balances` parameter.
+pub(crate) trait ArbitrageForCpmm<Balance, MarketId>
+where
+    MarketId: MaxEncodedLen,
+{
+    fn calc_total_spot_price(
+        &self,
+        balances: &BTreeMap<Asset<MarketId>, Balance>,
+    ) -> Result<Fixed, &'static str>;
+
+    fn calc_arbitrage_amount_mint_sell(
+        &self,
+        balances: &BTreeMap<Asset<MarketId>, Balance>,
+        max_iterations: usize,
+    ) -> Result<(Balance, usize), &'static str>;
+
+    fn calc_arbitrage_amount_buy_burn(
+        &self,
+        balances: &BTreeMap<Asset<MarketId>, Balance>,
+        max_iterations: usize,
+    ) -> Result<(Balance, usize), &'static str>;
+}
+
+impl<Balance, MarketId> ArbitrageForCpmm<Balance, MarketId> for Pool<Balance, MarketId>
+where
+    Balance: AtLeast32BitUnsigned + Copy,
+    MarketId: MaxEncodedLen + AtLeast32Bit + Copy,
+    Pool<Balance, MarketId>: ArbitrageForCpmmHelper<Balance, MarketId>,
+{
+    fn calc_total_spot_price(
+        &self,
+        balances: &BTreeMap<Asset<MarketId>, Balance>,
+    ) -> Result<Fixed, &'static str> {
+        let weights = self.weights.as_ref().ok_or("Unexpectedly found no weights in pool.")?;
+        let balance_in = balances
+            .get(&self.base_asset)
+            .cloned()
+            .ok_or("Base asset balance missing")?
+            .saturated_into();
+        let weight_in = weights
+            .get(&self.base_asset)
+            .cloned()
+            .ok_or("Unexpectedly found no weight for base asset")?;
+        let mut result: Fixed = 0;
+        for asset in self.assets.iter().filter(|a| **a != self.base_asset) {
+            let balance_out: Fixed = balances
+                .get(asset)
+                .cloned()
+                .ok_or("Asset balance missing from BTreeMap")?
+                .saturated_into();
+            let weight_out = weights
+                .get(asset)
+                .cloned()
+                .ok_or("Unexpectedly found no weight for outcome asset.")?;
+            // We're deliberately _not_ using the pool's swap fee!
+            result = result.saturating_add(calc_spot_price(
+                balance_in,
+                weight_in,
+                balance_out,
+                weight_out,
+                0,
+            )?);
```

Comment:
You want the total spot price, so the sum of all outcome_token prices (denominated in ZTG). Shouldn't be the denominator ZTG then?

From swaps/src/math:

```rust
let denom = bdiv(asset_balance_out, asset_weight_out)?;
```

But you are using the outcome asset balances as the denominator?

Reply:
_Denominator_ refers to the denominator of the fractiion in the formula for the spot price of swapping A_i for A_o (i for "in", o for "out"):

```
spot_price_io = (b_i / w_i) / (b_o / w_o)
```

It has nothing to do with the price being denominated in ZTG.

Reply:
"denominated in ZTG" just means the denominator of the fraction for me. For example: when I hear "price denominated in USD" I instantly think about the division (value / USD like for example KSMUSDT means KSM / USD). 

I mean when you have the following:

`swaps/src/math.rs` in `calc_spot_price`:
```rust
pub fn calc_spot_price(
    asset_balance_in: u128,
    asset_weight_in: u128,
    asset_balance_out: u128,
    asset_weight_out: u128,
    swap_fee: u128,
) -> Result<u128, DispatchError> {
    let numer = bdiv(asset_balance_in, asset_weight_in)?;
    let denom = bdiv(asset_balance_out, asset_weight_out)?;
    let ratio = bdiv(numer, denom)?;
    let scale = bdiv(BASE, BASE.check_sub_rslt(&swap_fee)?)?;
    let spot_price = bmul(ratio, scale);
    spot_price
}
```

Above is the division: `numer / denom`.

And the sum here:

```rust
result = result.saturating_add(calc_spot_price(
    balance_in_base_asset,
    weight_in_base_asset,
    balance_out_outcome_asset,
    weight_out_outcome_asset,
    0,
)?);
```

So, it's `base_asset / outcome_asset`. But I would expect the following:

`total_spot_price = (outcome_asset_A / base_asset) + (outcome_asset_B / base_asset) + (outcome_asset_C / base_asset)`

But I derive the following implementation at the moment:

`total_spot_price = (base_asset / outcome_asset_A) + (base_asset / outcome_asset_B) + (base_asset / outcome_asset_C)`

Reply:
> I expected that, because of my assumption, that when I value something in the base asset, then the base asset should be the denominator of a fraction (numerator / denominator).

Oh, I see. Well, the unit of the spot price is actually A_i / A_o. I always thought that was confusing, but if you have a currency pair KSM/USD, then the spot price is measured in USD/KSM.